aiida_vasp/calcs/workchains/converge.py

```diff
@@ -0,0 +1,1369 @@
+
+        self.ctx.workchains = []
+        self.ctx.pw_workchains = []
+        self.ctx.kpoints_workchains = []
+        self.ctx.running_pw = False
+        self.ctx.running_kpoints = False
+        self.ctx.replace_nodes = True
+
+        return
+
+    def _init_converge_context(self):
+        """Initialize the converge part of the context."""
+        self.ctx.converge = AttributeDict()
+        self.ctx.converge.settings = AttributeDict()
+        self._init_pw_context()
+        self._init_kpoints_context()
+        self._init_param_context()
+        self._check_context()
+
+        return
+
+    def _init_pw_context(self):
+        """
+        Initialize plane wave cutoff variables and store in context.
+
+        Parameters
+        ----------
+        encut : float, optional
+            The plane wave cutoff in eV. If not set, convergence tests on this
+            parameter are intitiated.
+        encut_start : float, optional
+            The start value of `encut` during convergence tests.
+        encut_step : float, optional
+            The `encut` step size in eV.
+        encut_samples : int
+            The number of `encut` steps to perform.
+        """
+        converge_dict = self.ctx.converge.settings
+        converge_dict['encut'] = None
+        converge_dict['encut_start'] = None
+        converge_dict['encut_step'] = None
+        converge_dict['encut_samples'] = None
+        if 'settings' in self.inputs:
+            settings = self.inputs.settings.get_dict()
+            if 'converge' in settings:
+                converge = settings.get('converge')
+                converge_dict['encut'] = converge.get('pw_energy_cutoff')
+                converge_dict['encut_start'] = converge.get('pw_energy_cutoff_start')
+                converge_dict['encut_step'] = converge.get('pw_energy_cutoff_step')
+                converge_dict['encut_samples'] = converge.get('pw_energy_cutoff_samples')
+        # Also check if encut is supplied in the incar input, this takes presence over
+        # the encut supplied in settings
+        if self.inputs.incar:
+            incar_dict = self.inputs.incar.get_dict()
+            try:
+                encut = incar_dict['encut']
+                converge_dict['encut'] = encut
+            except KeyError:
+                pass
+
+        return
+
+    def _init_kpoints_context(self):
+        """
+        Initialize the k-point grid variables and store in context.
+
+        Parameters
+        ----------
+        kgrid : (3) list int, optional
+            The k-point grid sampling along each axis, e.g. [3,3,3]. If not set,
+            convergence test on the k-point grid is initiated.
+        k_step : float, optional
+            The k-point step size.
+        k_samples : int, optional
+            The number of k-point grid refinements to perform with
+            the stepping of `k_step`.
+        """
+        converge_dict = self.ctx.converge.settings
```

Comment:
Another good place to use defaults for `dict.get` to do assignment and initialization in one go.

Reply:
This I do not understand. Think I only do this in order to not have to write `self.ctx.converge.settings`.

Reply:
Right. I put in an issue for that so I will leave these improvements for now.